### ISSUE TITLE
Low Success Rate( PickCube) : UR5e Task Stuck at Grasping, Never Places (Reward Plateaus at ~2.5/5.0)

### ISSUE THREAD

**Issue by saifi-shahrukh** [haosulab/ManiSkill#1387]
**Summary**
I'm training a custom UR5e + Robotiq Hand-E pick-and-place task based on PickCube, and the policy consistently learns to reach and grasp, but never places the object at the goal.

Training permanently plateaus at ~2.5 reward (50% of max) even after 50M+ training steps across multiple runs.

This behavior is consistent across different PD gains, PPO configurations, and episode lengths.

I'm planning sim-to-real deployment, and I also plan to test different observation spaces and controllers on the same setup, so resolving this issue is important.

<img width="1855" height="833" alt="Image" src="https://github.com/user-attachments/assets/6d56dea7-32e8-46f4-9e0c-7244d4813f64" />

**Attachments**

I am  providing the following as a zip file :

ur5e.urdf, agent config, environment code, training script, training curves

**Please let me know what would help   @StoneT2000 **

[ur5e.zip](https://github.com/user-attachments/files/25318847/ur5e.zip)

**Comment by saifi-shahrukh** [comment]
Hi,

I’m launching training via a wrapper script that builds the following command:

python baselines/ppo/ppo.py \
  --env_id PickCubeUR5e-v1 \
  --num_envs 4096 \
  --num_eval_envs 16 \
  --num_eval_steps 150 \
  --obs_mode state \
  --reward_mode dense \
  --control_mode pd_joint_delta_pos \
  --total_timesteps 50000000 \
  --update_epochs 8 \
  --num_minibatches 64 \
  --num_steps 150 \
  --track \
  --wandb_project_name rlzoo-project \
  --exp_name ur5e_pd_joint_delta_pos_state_dense_v1 \
  --render_mode rgb_array


-- *i am getting plateau at the placement reward and stuck in local minima*
--attached video after 25M steps training. 

https://github.com/user-attachments/assets/5f7248d6-53d0-4fb6-aae4-aaffd0629e6f

**Comment by StoneT2000** [comment]
Some initial RL env tuning recommendations
- Max environment episode steps of 500 for training is way too long. I recommend tuning it down to maybe even 100 steps or lower (I usually set it to something like 1.2-1.5x the typical # of steps it should take for the robot to solve the task). With very long episode horizons, it takes a long time before an environment is reset, leading to a lot of rollouts having very similar "stale" data (you can already observe it happening in the eval video which is 150 steps, for about 2 seconds the robot is doing some thing different, the rest of the time it barely makes much difference). This is critical for improving rollout buffer diversity
- Once you reduce the max env episode steps, reduce num_eval_steps and num_steps down to a number <= max env episode steps. For optimal rollout buffer data distributions technically num_steps should equal max env episode steps, but you often can get away with a smaller number and solve the task faster
- You can sweep over the gamma parameter. Our default is 0.8 which works well for these dense rewards and short skill horizon tasks, but it's not an exact science, you can try 0.9, 0.95, 0.99 (some common values used in RL).

And finally as a sanity check I recommend using a motion planner or some teleop to verify the task is solvable (e.g. the robot can reach the goal). Perhaps the robot is not defined well (e.g. wrong joint limits).

**Comment by saifi-shahrukh** [comment]
Hi,

Thank you for the suggestions. I have gone through the recommendations and wanted to share my setup and clarify a few points.
*URDF and Robot Definition*
-I am using the official UR5e URDF asset provided with ManiSkill, without modifying the arm joint limits. The joint limits in my URDF are:
- Shoulder pan, lift, wrist joints: ±2π
- Elbow joint: ±π
- Gripper fingers: 0 to 0.025 m (Robotiq Hand-E specification)

The URDF is attached, along with my:
- ur5e_robotiq_wristcam.py robot file
- pick_cube.py environment file

These are directly based on the ManiSkill asset structure.

The robot base is loaded at the world origin, and the workbench is defined as a fixed table with collision and visual geometry. The cube spawn height is aligned with the table surface:
*xyz[:, 2] = self.table_surface_z + self.cube_half_size*

So the object is correctly placed on the table and reachable by the robot.
I also verified that the robot can reach the cube and goal positions using the predefined pick_ready keyframe.
Aslo attached the workspace setup.

<img width="1879" height="1106" alt="Image" src="https://github.com/user-attachments/assets/bf82b289-04a5-405b-9f2d-d22eb43a5e6f" />

PPO Training Configuration Changes
Based on your recommendation, I reduced the episode horizon and rollout length.
My current PPO training configuration is:
max_episode_steps = 100
num_eval_steps = 100
gamma = 0.90
num_steps = 16 (reduced from 200)
total_timesteps = 50M on Ryzen 9 9950X + RTX 5070 Ti with 16 VRAM
This significantly improves rollout diversity and training stability compared to longer horizons.

<img width="1734" height="1091" alt="Image" src="https://github.com/user-attachments/assets/8e210646-e230-4850-955d-d47104d9c6b1" />
-----------------------------------------------------------------------------------------------------------------------------------------------------
Other DOUBT : regarding PPO RGB training with wrist camera

My setup uses only the wrist-mounted camera, similar to the ManiSkill wristcam environments. # ATTACHED IMAGE
*Question 1*

Is it necessary that the gripper fingers must be visible in the camera image for successful training?

If finger visibility is optional, how does the robot learn:
to localize itself relative to the cube?
to search for the cube on the table?

*Question 2*

Is there a recommended way to include multiple stacked frames (e.g., last 3-4 frames) in the observation space when using ppo_rgb.py?

This would help provide temporal information such as motion cues.
For example, something like:
obs = [frame_t, frame_t-1, frame_t-2, frame_t-3]


Is frame stacking already supported in ManiSkill PPO RGB training as with single rgb image and with not gripper finger int eh image frame .My environment is struggle to learn ?

Thank you again for your help. I have attached:

URDF
robot file
environment file
training config

Please let me know if you see anything incorrect in the setup.

[pick_cube.py](https://github.com/user-attachments/files/25419853/pick_cube.py)

[train_headless_fast.py](https://github.com/user-attachments/files/25419858/train_headless_fast.py)

[train_headless_rgb.py](https://github.com/user-attachments/files/25419861/train_headless_rgb.py)

[assets.zip](https://github.com/user-attachments/files/25419882/assets.zip)

[ur5e_robotiq_wristcam.py](https://github.com/user-attachments/files/25419887/ur5e_robotiq_wristcam.py)

**Comment by StoneT2000** [comment]
Great that the rl suggestions I made worked.

For frame stacking there is a frame stacking wrapper we provide, you can find it on the documentation on wrappers

For what image data to include, there isn't an exact science. The rule of thumb is if you are looking at the images yourself would you know what to do next.

**Comment by saifi-shahrukh** [comment]
### Massive performance gap between State vs. RGB policies on UR5e PickCube (State succeeds, RGB flatlines)

Hi again,

Thanks to your previous suggestions, 
I successfully tuned the environment and solved the task using state-based observations. The URDF, joint limits, and dense reward logic are all verified and working perfectly.

However, when transitioning to the visual policy `(ppo_rgb.py)` using the exact same control mode `(pd_joint_delta_pos)` and enabling `include_state=True,` the agent completely fails to learn. I have run extensive hyperparameter sweeps on my RTX 5070 Ti (16GB VRAM), and the visual runs flatline near 0 success.

I am hoping you can help me spot the bottleneck in my visual setup. 
_Below is the detailed comparison of the working vs. failing runs._

### 1. The Baseline: State-Based (Highly Successful)
Using ppo_fast.py, the UR5e solves the task quickly (completing in ~70-75 steps max even with 0.03 delta step). The robot learns to reach, grasp, and place flawlessly.

```
python baselines/ppo/ppo_fast.py \
  --env_id PickCubeUR5e-v1 --seed 1 --num_envs 4096 --num_eval_envs 16 \
  --num_eval_steps 150 --obs_mode state --reward_mode dense \
  --control_mode pd_joint_delta_pos --total_timesteps 50000000 \
  --update_epochs 8 --num_minibatches 32 --gamma 0.97 --num_steps 16 \
  --track --wandb_project_name rlzoo-project \
  --exp_name ur5e_pd_joint_delta_pos_state_dense__fresh_v1 --cudagraphs
```
Result(See attached Image 1): Rapid convergence, high success rate.

<img width="1848" height="925" alt="Image" src="https://github.com/user-attachments/assets/7d441a2f-519e-40f7-9b16-c1216e9f6a41" />

### 2. The Failing Run: RGB + include_state (Matched with your reference config but with 0.05 delta step)
After trying our different variations, at last, I attempted to replicate the exact math of your ppo-PickCube-v1-rgb-4796-walltime_efficient reference run. To fit my 16GB VRAM, I halved the envs and doubled the rollout steps to maintain the exact 16,384 batch size. I also used normalized_dense and gamma 0.8 as recommended.

Command:
```
python baselines/ppo/ppo_rgb.py \
  --env_id PickCubeUR5e-v1 --num_envs 512 --num_eval_envs 8 \
  --num_eval_steps 50 --obs_mode rgb --reward_mode normalized_dense \
  --control_mode pd_joint_delta_pos --total_timesteps 50000000 \
  --update_epochs 8 --num_minibatches 32 \
  --num_steps 32 --gamma 0.80 --track \
  --exp_name ur5e_visual_pd_joint_delta_pos_rgb_normalized_dense_v1
```
Result (See attached Image 2): Total flatline. success_once stays at 0. The policy loss and value loss exhibit high variance, and train/episode_len stays pinned at 50, indicating the robot never completes the task early.

<img width="1847" height="923" alt="Image" src="https://github.com/user-attachments/assets/326676c4-b3b1-439e-a4b7-76b831a495ec" />

### 3. The Failing Run 2: RGB + Frame Stacking +  include_state
Thinking the agent lacked temporal context/motion cues, I tried using the frame stacking wrapper alongside the raw dense reward and a longer horizon.

Command:
```
python baselines/ppo/ppo_rgb.py \
  --env_id PickCubeUR5e-v1 --num_envs 64 --num_eval_envs 8 \
  --num_eval_steps 100 --obs_mode rgb --reward_mode dense \
  --control_mode pd_joint_delta_pos --num_stack 3 \
  --total_timesteps 50000000 --update_epochs 8 --num_minibatches 16 \
  --num_steps 50 --gamma 0.99 --track \
  --render_mode rgb_array --no-partial_reset
```
Result (See attached Image  3): Slight spike in success_once to ~0.125, but immediately crashes and flatlines.

<img width="1846" height="929" alt="Image" src="https://github.com/user-attachments/assets/c23fa73c-c55f-4114-95f6-f7a5a78cac0f" />

### Key Questions for Diagnosis:
**1. Action Delta Step:** What changes in learning dynamics should I expect if I keep the delta step in the action space lower compared to the baseline? I have observed differences in my runs when tweaking this and want to ensure I am using the optimal scale for visual RL.

**2. Control Mode Selection**: When starting with a vision-based policy, do you recommend using `pd_joint_delta_pos` or `pd_ee_delta_pos` to successfully map pixels to actions?

**3. Visual + State Fusion**: I am exclusively using the wrist-cam for training, and as shown in my earlier images, both the gripper and the cube are clearly visible in the frame. Given this clear visibility alongside the proprioceptive state data (`include_state=True`), why might the success rate still remain so low? Can you suggest or provide 2-3 command runs/configurations for testing this specific setup?

**4. Pure Vision (`no-include-state`) Baseline**: Have you tried training the RGB vision policy for the Panda without the state data (`--no-include-state`)? What are the practical chances of solving the task this way, and what specific parameter adjustments would you recommend for that setup given my 16GB VRAM hardware constraints?

Your advice would be greatly appreciated 🙏!

**Comment by StoneT2000** [comment]
- You might need to tune the camera settings. The default PickCube env uses a third-eye camera and no wrist camera. The harder tasks like StackCube use third eye camera and a wrist camera. Try adding a third-eye camera and see how it performs. 
- Pure vision policy is not recommended. It slows down training

**Comment by saifi-shahrukh** [comment]
Hi,

I trained PickCubeUR5e-v1 using both a third-person camera and a wrist camera, following scripts very similar to the ManiSkill examples. 
<img width="2104" height="1276" alt="Image" src="https://github.com/user-attachments/assets/69c4835d-1fcc-465b-a395-e8feff5cc4d5" />
General Training Setup:
```
Environment:

env_id = "PickCubeUR5e-v1"

control_mode = "pd_joint_delta_pos"

sim_backend = "physx_cuda"

num_envs = 256

num_steps = 50

num_minibatches = 16

total_timesteps = 50_000_000

Key PPO parameters:

learning_rate = 3e-4

gamma = 0.8

gae_lambda = 0.9

clip_coef = 0.2

update_epochs = 8
```

After ~20M timesteps, I reach around 80% success rate with these configuration.

_**Observations**_

1. num_envs, num_steps, and num_minibatches significantly affect learning stability and final performance. Changing these values noticeably alters convergence speed and success rate.
2. Gripper visibility matters a lot. When the fingers are visible in the wrist camera view, grasping becomes noticeably easier for the policy.

_**Questions**_

**1. Effect of Smaller Action Deltas**

If I reduce the magnitude of joint delta actions (i.e., make the delta position increments smaller), how much should I expect this to affect training?

**2.  Domain Randomization for Sim2Real**

There are noticeable visual differences between simulator images and the real setup.

What aspects should be randomized for better transfer?

For example:
- Lighting intensity and direction?
- Texture and color of objects?
- Table surface appearance?
- Camera noise / blur?
- Camera intrinsics and slight pose perturbations?

Is there a recommended minimal set of randomizations that usually helps most for wrist-camera manipulation?

**3. Sim2Real Transfer**

I trained the policy in simulation and attempted deployment on the real robot using:

mani_skill.agents.base_real_agent and provided sim2real wrapper

However, I’m struggling to get it working reliably on hardware.

_Questions:_

Is there a recommended deployment pipeline for policies trained in PickCube-v1?

Any guidance on a robust sim2real workflow would be very helpful.

**experimented with Wrist camera with gripper visible**
https://github.com/user-attachments/assets/1e288fda-c76c-444b-8717-af6c70360d30

<img width="1737" height="1025" alt="Image" src="https://github.com/user-attachments/assets/603fd476-95be-49b5-83c0-d18630da462a" />

**experimented with Wrist camera without gripper visible**
https://github.com/user-attachments/assets/298c69e4-bb51-4562-99fa-25d52354d5e2

<img width="1729" height="1020" alt="Image" src="https://github.com/user-attachments/assets/0e2db26c-a75f-474b-ba46-5cd7ec620b64" />

**Comment by StoneT2000** [comment]
I'm not sure what you mean by you can't get it to work on real hardware. Did you define a real agent interface for the real robot in python?

As for sim2real, there are two repos I recommend following to learn more.

https://github.com/aalmuzairee/squint (supports wrist cameras and has nice techniques for visual sim2real)

https://github.com/StoneT2000/lerobot-sim2real - older lerobot based sim2real project.

for smaller action deltas, that's perfectly fine (and might lead to lower sim2real gap), but will lead to longer task horizons and so you need to adjust max episode steps and maybe exploration accordingly.

for visual randomizations, the squint repo has some examples. Other papers have also tried all kinds of randomizations, there is no clear answer however.
